File: TelerikUIExamples/TelerikUIExamples/GaugeRanges.m
//
//  GaugeRanges.m
//  TelerikUIExamples
//
//  Copyright © 2015 г. Telerik. All rights reserved.
//

#import "GaugeRanges.h"
#import <TelerikUI/TelerikUI.h>

@interface GaugeRanges ()

@property (nonatomic, strong) TKLinearGauge* linearGauge;
@property (nonatomic, strong) TKRadialGauge* radialGauge;

@end

@implementation GaugeRanges
{
    NSArray* _redColors;
    NSArray* _greenColors;
    NSArray* _blueColors;
    NSArray* _redValues;
    NSArray* _blueValues;
    NSArray* _greenValues;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _redColors =   @[[UIColor colorWithRed:0.96 green:0.80 blue:0.80 alpha:1.0],
                     [UIColor colorWithRed:0.80 green:0.00 blue:0.00 alpha:1.0],
                     [UIColor colorWithRed:0.88 green:0.40 blue:0.40 alpha:1.0]];
    
    _greenColors = @[[UIColor colorWithRed:0.85 green:0.92 blue:0.83 alpha:1.0],
                     [UIColor colorWithRed:0.71 green:0.84 blue:0.66 alpha:1.0],
                     [UIColor colorWithRed:0.58 green:0.77 blue:0.49 alpha:1.0]];
    
    _blueColors  = @[[UIColor colorWithRed:0.62 green:0.77 blue:0.91 alpha:1.0] ,
                     [UIColor colorWithRed:0.44 green:0.66 blue:0.86 alpha:1.0],
                     [UIColor colorWithRed:0.11 green:0.27 blue:0.53 alpha:1.0]];
    
    _redValues = @[@(0), @(20), @(40), @(100)];
    _greenValues = @[@(0), @(40), @(80), @(100)];
    _blueValues = @[@(0), @(10), @(90), @(100)];
    
    NSArray* legendStrings = @[@"FATS", @"CARBS", @"PROTEINS"];
    NSArray* legendColors = @[[UIColor colorWithRed:0.96 green:0.80 blue:0.80 alpha:1.0],
                              [UIColor colorWithRed:0.85 green:0.92 blue:0.83 alpha:1.0],
                              [UIColor colorWithRed:0.62 green:0.77 blue:0.91 alpha:1.0]];
    
    for (int i=0; i < 3; i++) {
        TKView *view = [[TKView alloc] initWithFrame:CGRectMake(20, 80 + i*25, 22, 22)];
        view.fill = [TKSolidFill solidFillWithColor:legendColors[i]];
        view.shape = [TKPredefinedShape shapeWithType:TKShapeTypeCircle andSize:CGSizeZero];
        [self.view addSubview:view];
        
        UILabel* label = [[UILabel alloc] initWithFrame:CGRectMake(50, 80 + i * 25, CGRectGetMaxX(self.view.frame) - 60, 20)];
        label.text = legendStrings[i];
        label.font = [UIFont fontWithName:@"HelveticaNeue-Light" size:15];
        label.textColor = [UIColor colorWithRed:0.2 green:0.2 blue:0.2 alpha:1];
        [self.view addSubview:label];
    }

    [self setLinearGauge];
    [self setRadialGauge];
}

- (void)setLinearGauge
{
    self.linearGauge = [[TKLinearGauge alloc] initWithFrame:CGRectZero];
    [self.view addSubview:self.linearGauge];
    
    TKGaugeLinearScale* scale = [[TKGaugeLinearScale alloc] init];
    [self.linearGauge addScale:scale];
  
    for (int i = 0; i < 3; i++) {
        TKGaugeSegment* segment = [[TKGaugeSegment alloc] init];
        segment.range = [[TKRange alloc]initWithMinimum:_redValues[i] andMaximum:_redValues[i+1]];
        segment.fill = [TKSolidFill solidFillWithColor:_redColors[i]];
        segment.location = 0.55;
        segment.cap = TKGaugeSegmentCapRound;
        [scale addSegment:segment];
    }
  
    for (int i = 0; i < 3; i++) {
        TKGaugeSegment* segment = [[TKGaugeSegment alloc] init];
        segment.range = [[TKRange alloc]initWithMinimum:_greenValues[i] andMaximum:_greenValues[i+1]];
        segment.fill = [TKSolidFill solidFillWithColor:_greenColors[i]];
        segment.location = 0.67;
        segment.cap = TKGaugeSegmentCapRound;
        [scale addSegment:segment];
    }
    
    for (int i = 0; i < 3; i++) {
        TKGaugeSegment* segment = [[TKGaugeSegment alloc] init];
        segment.range = [[TKRange alloc]initWithMinimum:_blueValues[i] andMaximum:_blueValues[i+1]];
        segment.fill = [TKSolidFill solidFillWithColor:_blueColors[i]];
        segment.location = 0.79;
        segment.cap = TKGaugeSegmentCapRound;
        [scale addSegment:segment];
    }
    
    NSArray* lengths = @[@.55, @.67, @.79];
    
    for (int i = 0; i < 3 ; i++) {
        TKGaugeNeedle* needle = [TKGaugeNeedle new];
        needle.value = arc4random_uniform(100);
        needle.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
        needle.length = 1 - [lengths[i] doubleValue];
        [scale addIndicator:needle];
    }
}

- (void)setRadialGauge
{
    self.radialGauge = [[TKRadialGauge alloc] initWithFrame:CGRectZero];
    [self.view addSubview:self.radialGauge];
    
    TKGaugeRadialScale* scale = [[TKGaugeRadialScale alloc] init];
    scale.radius = 0.76;
    scale.ticks.position = TKGaugeTicksPositionOuter;
    scale.labels.position = TKGaugeTicksPositionOuter;
    scale.ticks.offset = 0.05;
    scale.labels.offset = 0.15;
    [self.radialGauge addScale:scale];
    
      for (int i = 0; i < 3; i++) {
        TKGaugeSegment* segment = [[TKGaugeSegment alloc] init];
        segment.range = [[TKRange alloc]initWithMinimum:_redValues[i] andMaximum:_redValues[i+1]];
        segment.fill = [TKSolidFill solidFillWithColor:_redColors[i]];
        segment.cap = TKGaugeSegmentCapRound;
        segment.location = .76;
        [scale addSegment:segment];
    }
    
    for (int i = 0; i < 3; i++) {
        TKGaugeSegment* segment = [[TKGaugeSegment alloc] init];
        segment.range = [[TKRange alloc]initWithMinimum:_greenValues[i] andMaximum:_greenValues[i+1]];
        segment.fill = [TKSolidFill solidFillWithColor:_greenColors[i]];
        segment.location = .64;
        segment.cap = TKGaugeSegmentCapRound;
        [scale addSegment:segment];
    }
    
    for (int i = 0; i < 3; i++) {
        TKGaugeSegment* segment = [[TKGaugeSegment alloc] init];
        segment.range = [[TKRange alloc]initWithMinimum:_blueValues[i] andMaximum:_blueValues[i+1]];
        segment.fill = [TKSolidFill solidFillWithColor:_blueColors[i]];
        segment.location = .52;
        segment.cap = TKGaugeSegmentCapRound;
        [scale addSegment:segment];
    }
    
    NSArray* lengths = @[@.76, @.64, @.52];
    
    for (int i = 0; i < 3 ; i++) {
        TKGaugeNeedle* needle = [TKGaugeNeedle new];
        needle.value = arc4random_uniform(100);
        needle.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
        needle.length = [lengths[i] doubleValue];
        [scale addIndicator:needle];
    }
}

#pragma mark - Layout

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    CGRect bounds = self.exampleBounds;
    CGSize size = self.view.bounds.size;
    CGFloat offset = 20;
    CGFloat linearHeight = 135;
    
    if (UIDeviceOrientationIsLandscape([UIDevice currentDevice].orientation)) {
        CGFloat width = (CGRectGetWidth(bounds)- offset* 3.)/2.;
        CGFloat height = CGRectGetHeight(bounds) - offset*2.;
        _linearGauge.frame = CGRectMake(offset, offset+ (size.height -linearHeight)/2 , width, linearHeight);
        _radialGauge.frame = CGRectMake(2*offset + width, offset + (size.height - height)/2, width, height);
    } else {
        CGFloat width = CGRectGetWidth(bounds) - offset*2.;
        CGFloat height = (CGRectGetHeight(bounds)- offset* 3.)/2.;
        _linearGauge.frame = CGRectMake(offset, (offset + height) /2., width, linearHeight);
        _radialGauge.frame = CGRectMake(offset, (size.height)/2., width, height + offset*2);
    }
}

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/GaugeRanges.swift
//
//  GaugeRanges.swift
//  TelerikUIExamplesInSwift
//
//  Copyright © 2015 Telerik. All rights reserved.
//

import UIKit

class GaugeRanges: TKExamplesExampleViewController {
    
    let linearGauge = TKLinearGauge()
    let radialGauge = TKRadialGauge()

    let redColors   = [ UIColor(red: 0.96, green: 0.80, blue: 0.80, alpha: 1.00),
                        UIColor(red: 0.80, green: 0.00, blue: 0.00, alpha: 1.00),
                        UIColor(red: 0.88, green: 0.40, blue: 0.40, alpha: 1.00) ]
    
    let greenColors = [ UIColor(red: 0.85, green: 0.92, blue: 0.83, alpha: 1.00),
                        UIColor(red: 0.71, green: 0.84, blue: 0.66, alpha: 1.00),
                        UIColor(red: 0.58, green: 0.77, blue: 0.49, alpha: 1.00) ]
    
    let blueColors  = [ UIColor(red: 0.62, green: 0.77, blue: 0.91, alpha: 1.00),
                        UIColor(red: 0.44, green: 0.66, blue: 0.86, alpha: 1.00),
                        UIColor(red: 0.11, green: 0.27, blue: 0.53, alpha: 1.00) ]

    let redValues = [ 0, 20, 40, 100 ]
    let greenValues = [ 0, 40, 80, 100 ]
    let blueValues = [ 0, 10, 90, 100 ]
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let legendStrings = [ "FATS", "CARBS", "PROTEINS" ]
        let legendColors = [ UIColor(red: 0.96, green: 0.80, blue: 0.80, alpha: 1.00),
                             UIColor(red: 0.85, green: 0.92, blue: 0.83, alpha: 1.00),
                             UIColor(red: 0.62, green: 0.77, blue: 0.91, alpha: 1.00)]

        for i in 0..<3 {
            let view = TKView(frame: CGRectMake(20, CGFloat(80 + i*25), 22, 22))
            view.fill = TKSolidFill(color: legendColors[i])
            view.shape = TKPredefinedShape(type: TKShapeType.Circle, andSize: CGSizeZero)
            self.view.addSubview(view)
            
            let label = UILabel(frame: CGRectMake(50, CGFloat(80 + i*25), CGRectGetMaxX(self.view.frame) - 60, 20))
            label.text = legendStrings[i]
            label.font = UIFont(name: "HelveticaNeue-Light", size: 15)
            label.textColor = UIColor(red: 0.2, green: 0.2, blue: 0.2, alpha: 1)
            self.view.addSubview(label)
        }
        
        setLinearGauge()
        setRadialGauge()
    }
    
    func setLinearGauge() {
        
        self.view.addSubview(self.linearGauge)
        
        let scale = TKGaugeLinearScale()
        self.linearGauge.addScale(scale)
        
        for i in 0..<3 {
            let segment = TKGaugeSegment(minimum: self.redValues[i], maximum: self.redValues[i+1])
            segment.fill = TKSolidFill(color: self.redColors[i])
            segment.location = 0.55
            segment.cap = TKGaugeSegmentCap.Round
            scale.addSegment(segment)
        }
        
        for i in 0..<3 {
            let segment = TKGaugeSegment(minimum: self.greenValues[i], maximum: self.greenValues[i+1])
            segment.fill = TKSolidFill(color: self.greenColors[i])
            segment.location = 0.67
            segment.cap = TKGaugeSegmentCap.Round
            scale.addSegment(segment)
        }
        
        for i in 0..<3 {
            let segment = TKGaugeSegment(minimum: self.blueValues[i], maximum: self.blueValues[i+1])
            segment.fill = TKSolidFill(color: self.blueColors[i])
            segment.location = 0.79
            segment.cap = TKGaugeSegmentCap.Round
            scale.addSegment(segment)
        }
        
        let lengths = [0.55, 0.67, 0.79]
        for i in 0..<3 {
            let needle = TKGaugeNeedle(value: CGFloat(arc4random_uniform(100)))
            needle.fill = TKSolidFill(color: UIColor.grayColor())
            needle.length = CGFloat(1 - lengths[i])
            scale.addIndicator(needle)
            
        }
    }
    
    func setRadialGauge() {

        self.view.addSubview(self.radialGauge)
        
        let scale = TKGaugeRadialScale()
        scale.radius = 0.76
        scale.ticks.position = TKGaugeTicksPosition.Outer
        scale.labels.position = TKGaugeLabelsPosition.Outer
        scale.labels.offset = 0.15
        scale.ticks.offset = 0.05
        self.radialGauge.addScale(scale)
        
        for i in 0..<3 {
            let segment = TKGaugeSegment(minimum: self.redValues[i], maximum: self.redValues[i+1])
            segment.fill = TKSolidFill(color: self.redColors[i])
            segment.cap = TKGaugeSegmentCap.Round
            segment.location = 0.76
            scale.addSegment(segment)
        }
        
        for i in 0..<3 {
            let segment = TKGaugeSegment(minimum: self.greenValues[i], maximum: self.greenValues[i+1])
            segment.fill = TKSolidFill(color: self.greenColors[i])
            segment.cap = TKGaugeSegmentCap.Round
            segment.location = 0.64
            scale.addSegment(segment)
        }
        
        for i in 0..<3 {
            let segment = TKGaugeSegment(minimum: self.blueValues[i], maximum: self.blueValues[i+1])
            segment.fill = TKSolidFill(color: self.blueColors[i])
            segment.cap = TKGaugeSegmentCap.Round
            segment.location = 0.52
            scale.addSegment(segment)
        }
        
        let lengths = [0.76, 0.64, 0.52]
        for i in 0..<3 {
            let needle = TKGaugeNeedle(value: CGFloat(arc4random_uniform(100)))
            needle.fill = TKSolidFill(color: UIColor.grayColor())
            needle.length = CGFloat(lengths[i])
            scale.addIndicator(needle)
        }
    }

    //MARK: - Layout
    
    override func viewDidLayoutSubviews() {

        super.viewDidLayoutSubviews()
        
        let bounds = self.view.bounds
        let size = self.view.bounds.size
        let offset = CGFloat(20)
        let linearHeight = CGFloat(150)
        
        if(UIDeviceOrientationIsLandscape(UIDevice.currentDevice().orientation)) {
            let width = (CGRectGetWidth(bounds) - offset*3.0)/2.0
            let height = CGRectGetHeight(bounds) - offset*2.0
            self.linearGauge.frame = CGRectMake(offset, offset + (size.height - linearHeight)/2, width, linearHeight)
            self.radialGauge.frame = CGRectMake(2*offset + width, offset + (size.height - height)/2, width, height)
            
        }
        else {
            let width = CGRectGetWidth(bounds) - offset*2.0
            let height = (CGRectGetHeight(bounds) - offset*3.0)/2.0
            self.linearGauge.frame = CGRectMake(offset, (offset + height)/2.0, width, linearHeight)
            self.radialGauge.frame = CGRectMake(offset, (size.height + offset)/2.0, width, height)
        }
    }
}
